services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java

```diff
@@ -1931,6 +1932,50 @@ public void setStoreCurrentVersion(String clusterName, String storeName, int ver
             + "setting version on parent is not supported, since the version list could be different fabric by fabric");
   }
 
+  /**
+   * Set backup version as current version in all child regions.
+   */
+  @Override
+  public void rollbackToBackupVersion(String clusterName, String storeName) {
+    acquireAdminMessageLock(clusterName, storeName);
+    try {
+      // Check whether backup version is consistent in all child regions
+      Map<String, ControllerClient> controllerClientMap = getVeniceHelixAdmin().getControllerClientMap(clusterName);
+      Set<Integer> backupVersionSet = new HashSet<>();
+      controllerClientMap.forEach((regionName, cc) -> {
+        StoreResponse storeResponse = cc.getStore(storeName);
+        if (storeResponse.isError()) {
+          throw new VeniceException(storeResponse.getError() + " in region " + regionName);
+        }
+        if (!storeResponse.getStore().isEnableStoreWrites()) {
+          throw new VeniceException("Unable to rollback since store writeability is false");
+        }
```

Comment:
Wondering is it possible to move the rest calls out of the store level lock? The default timeout value for the request is 600 seconds, so we could be blocked here by 10 min while holding this lock in case of network issue. If you feel this is a bit risky, we can improve this later.

Reply:
Right. Including the controller client calls inside the lock might end up blocking the store's other operations for some time when there's network issue. Since the lock is store-level, other stores' operations won't be blocked.

Moving the calls out of the lock has another problem though: when a store is pushed and rolled back at the same time, even though rollback request succeeds, the actual version rollback in child regions might fail as backup version might be deleted. For example, a store has backup version v1 and current version v2. Rollback request comes in and finds out the backup version is v1. Meanwhile, a new push request for the store arrives, acquires the lock, and sends the admin message to add v3 and delete v1 (because early backup deletion is enabled). Then rollback request acquires the lock, send the admin message to set version to v1, and responds success. The set version admin message execution in child controller will always fail because v1 is gone.

Reply:
That's a valid problem. In my mind, one way to handle it could be when rollback request acquires the lock, it needs to check if the backupVersion that it gets is still valid (by valid I mean things that you mentioned e.g. version creation/deletion didn't actually happen during the period of getting the backupVersion and backupVersion is still the correct number). I wish this check only needs to access in-memory data, assuming that version creation have to go through parent controller as well. In case of detecting such operation conflict, we can simply choose to fail the rollback requests and ask it to retry later in favor of version creation operation. 

That being said, I do see that it looks like a pattern to acquire this store level lock first during a heavy operation for parent controller, and I don't know is if lock has caused us any trouble in the past. It could be the case that it is not too bad and hasn't been an issue for parent controller in the past, then maybe we can evaluate it later.

Reply:
Today checking whether the backup version is valid could not be done in parent controller. It needs to call child controller as well. Because parent controller does not maintain the status of versions (all versions are in started status). And it does not do early backup version deletion. It just simply keep the last 5 versions. In the above example, v1 will still exist in parent controller after a new push. That means, if parent controller wants to check if backup version is valid after acquiring the lock, it needs to call child controllers too inside the lock.

Reply:
Good discussion.
Can we do two things?
1. Send requests to all child colos in parallel.
2. Reduce the timeout from 10 mins to 1 min in the controller client being used by parent controller.

It a request will take more than 1 min, that must indicate a serious problem, and I am not sure whether 10mins will help considering  child controller is using store-level locking as well.

Reply:
Thanks for the suggestions! I updated the parent controller code to send requests in parallel. Also, the timeout of getStore request here is reduced from 10min to waitingTimeForConsumptionMs (default 30s). If the request to any child controller takes longer, rollback version request will fail and release the lock.

Reply:
Thanks Shuihui and Gaojie for the discussion! I think the alternative makes sense to me given what we have today.